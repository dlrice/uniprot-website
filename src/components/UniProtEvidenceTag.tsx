import React, { FC, Fragment } from 'react';
import { EvidenceTag, SwissProtIcon, TremblIcon } from 'franklin-sites';
import { html } from 'lit-html';
import {
  getEvidenceCodeData,
  EvidenceData,
} from '../model/types/EvidenceCodes';
import { Evidence } from '../model/types/modelTypes';
import { groupBy } from '../utils/utils';

export const UniProtEvidenceTagContent: FC<{
  evidenceData: EvidenceData;
  references: Evidence[] | undefined;
}> = ({ evidenceData, references }) => (
  <div>
    <h5>{evidenceData.label}</h5>
    {references &&
      references
        .filter((reference: Evidence) => reference.id && reference.source)
        .map((reference: Evidence) => (
          <div key={reference.id}>{`${reference.source}:${reference.id}`}</div>
        ))}
  </div>
);

const UniProtEvidenceTag: FC<{ evidences: Evidence[] }> = ({ evidences }) => {
  const evidenceMap: Map<string, Evidence[]> = groupBy(
    evidences,
    (evidence: Evidence) => evidence.evidenceCode
  );
  const evidenceTags = Array.from(evidenceMap.keys()).map(evidenceCode => {
    const evidenceData = getEvidenceCodeData(evidenceCode);
    if (!evidenceData) {
      return null;
    }
    const references = evidenceMap.get(evidenceCode);
    return (
      <EvidenceTag
        label={
          evidenceData.labelRender
            ? evidenceData.labelRender(references)
            : evidenceData.label
        }
        iconComponent={evidenceData.manual ? <SwissProtIcon /> : <TremblIcon />}
        className={
          evidenceData.manual ? 'svg-colour-reviewed' : 'svg-colour-unreviewed'
        }
        key={evidenceCode}
      >
        <UniProtEvidenceTagContent
          evidenceData={evidenceData}
          references={references}
        />
      </EvidenceTag>
    );
  });
  return <Fragment>{evidenceTags}</Fragment>;
};

export const UniProtProtvistaEvidenceTag = (
  evidences: Evidence[],
  callback: Function
) => {
  const evidenceMap: Map<string, Evidence[]> = groupBy(
    evidences,
    (evidence: Evidence) => evidence.evidenceCode
  );
  const evidenceTags = Array.from(evidenceMap.keys()).map(evidenceCode => {
    const evidenceData = getEvidenceCodeData(evidenceCode);
    if (!evidenceData) {
      return html``;
    }
    const references = evidenceMap.get(evidenceCode);
    return html`
      <span
        class="evidence-tag"
        @click=${() => callback(evidenceData, references)}
      >
        ${evidenceData.labelRender
          ? evidenceData.labelRender(references)
          : evidenceData.label}</span
      >
    `;
  });
  return evidenceTags;
};

export default UniProtEvidenceTag;
