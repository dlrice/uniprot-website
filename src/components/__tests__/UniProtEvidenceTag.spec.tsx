import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import UniProtEvidenceTag from '../UniProtEvidenceTag';

configure({ adapter: new Adapter() });

const evidence = {
  evidenceCode: 'ECO:0000255',
  source: 'PROSITE-ProRule',
  id: 'PRU10023',
};

describe('UniProtEvidenceTag components', () => {
  test('should render', () => {
    const wrapper = shallow(<UniProtEvidenceTag evidence={evidence} />);
    expect(wrapper.debug()).toMatchSnapshot();
  });
});
