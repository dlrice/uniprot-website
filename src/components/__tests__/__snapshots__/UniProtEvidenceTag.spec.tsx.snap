// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UniProtEvidenceTag components should render ProtVista evidence tag 1`] = `
Array [
  TemplateResult {
    "processor": DefaultTemplateProcessor {},
    "strings": Array [
      "
        <span
          class=",
      "
          @click=",
      "
        >
          ",
      "
          ",
      "</span
        >
      ",
    ],
    "type": "html",
    "values": Array [
      "evidence-tag svg-colour-unreviewed",
      [Function],
      [Function],
      "Imported",
    ],
  },
]
`;

exports[`UniProtEvidenceTag components should render automatic  1`] = `
<DocumentFragment>
  <span
    class="evidence-tag svg-colour-unreviewed"
    data-testid="evidence-tag-trigger"
    role="button"
    tabindex="0"
  >
    <svg
      fill="none"
      height="12"
      viewBox="0 0 10 15"
      width="12"
    >
      <path
        clip-rule="evenodd"
        d="M10 0H0v15h.051L5 10l4.949 5H10V0z"
        fill="#4E5A71"
        fill-rule="evenodd"
      />
    </svg>
    <span
      class="evidence-tag__label"
      title=""
    >
      Imported
    </span>
  </span>
  <div
    class="evidence-tag-content "
    data-testid="evidence-tag-content"
  >
    <div>
      <h5>
        Automatic assertion inferred from database entries
      </h5>
    </div>
  </div>
</DocumentFragment>
`;

exports[`UniProtEvidenceTag components should render automatic annotation 1`] = `
<DocumentFragment>
  <span
    class="evidence-tag svg-colour-reviewed"
    data-testid="evidence-tag-trigger"
    role="button"
    tabindex="0"
  >
    <svg
      fill="none"
      height="12"
      viewBox="0 0 10 15"
      width="12"
    >
      <path
        clip-rule="evenodd"
        d="M10 0H0v15h.051L5 10l4.949 5H10V0z"
        fill="#4E5A71"
        fill-rule="evenodd"
      />
    </svg>
    <span
      class="evidence-tag__label"
      title=""
    >
      1 automatic annotation
    </span>
  </span>
  <div
    class="evidence-tag-content "
    data-testid="evidence-tag-content"
  >
    <div>
      <h5>
        Automatic assertion according to rules
      </h5>
      <a
        href="//prosite.expasy.org/unirule/PRU10023"
      >
        PRU10023
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`UniProtEvidenceTag components should render publications count 1`] = `
<DocumentFragment>
  <span
    class="evidence-tag svg-colour-reviewed"
    data-testid="evidence-tag-trigger"
    role="button"
    tabindex="0"
  >
    <svg
      fill="none"
      height="12"
      viewBox="0 0 10 15"
      width="12"
    >
      <path
        clip-rule="evenodd"
        d="M10 0H0v15h.051L5 10l4.949 5H10V0z"
        fill="#4E5A71"
        fill-rule="evenodd"
      />
    </svg>
    <span
      class="evidence-tag__label"
      title=""
    >
      2 publications
    </span>
  </span>
  <div
    class="evidence-tag-content "
    data-testid="evidence-tag-content"
  >
    <div>
      <h5>
        Manual assertion based on experiment
      </h5>
    </div>
  </div>
</DocumentFragment>
`;
